lib/rpc/singleJsonRpcProvider.ts

```diff
@@ -0,0 +1,460 @@
+import { StaticJsonRpcProvider, TransactionRequest } from '@ethersproject/providers'
+import { CHAIN_IDS_TO_NAMES } from './chains'
+import { Config, DEFAULT_CONFIG } from './config'
+import Debug from 'debug'
+import { metric, MetricLoggerUnit } from '@uniswap/smart-order-router'
+import { ChainId } from '@uniswap/sdk-core'
+import {
+  BlockTag,
+  BlockWithTransactions,
+  Filter,
+  Log,
+  TransactionReceipt,
+  TransactionResponse,
+} from '@ethersproject/abstract-provider'
+import { BigNumber, BigNumberish } from '@ethersproject/bignumber'
+import { Network } from '@ethersproject/networks'
+import { Deferrable } from '@ethersproject/properties'
+const debug = Debug('SingleJsonRpcProvider')
+
+class PerfStat {
+  lastCallTimestampInMs: number = 0
+  lastCallSucceed: boolean = false
+  lastCallLatencyInMs: number = 0
+  timeWaitedBeforeLastCallInMs: number = 0
+}
+
+export default class SingleJsonRpcProvider extends StaticJsonRpcProvider {
+  // TODO(jie): Implement block-aligned cache
+  url: string
+  private healthScore
+  private isRecovering: boolean
+  private perf: PerfStat
+  private config: Config
+  private readonly metricPrefix: string
+
+  constructor(chainId: ChainId, url: string, config: Config = DEFAULT_CONFIG) {
+    super(url, { chainId, name: CHAIN_IDS_TO_NAMES[chainId] })
+    this.url = url
+    this.healthScore = 0
+    this.isRecovering = false
+    this.perf = new PerfStat()
+    this.config = config
+    this.metricPrefix = `RPC_${this.network.chainId}_${this.url}`
```

Comment:
Seeing this metrics prefix, I was worried about the API Key part of the RPC URL will get exported to the metrics. Checking with https://github.com/Uniswap/routing-api/pull/449/files#diff-e594d19771a20ac88f1ab3d2795c11801460d30a4d14aad748a4ba55987865f4R29-R30, that indeed seems like the case. 

I think we still want to call [deriveProviderName](https://github.com/Uniswap/routing-api/blob/main/lib/handlers/evm/provider/ProviderName.ts#L8) to have a more human-friendly as well as shortened metrics prefix.